aten/src/ATen/Allocator.h

```diff
@@ -3,14 +3,127 @@
 #include <memory>
 #include <stddef.h>
 
-#include "ATen/Retainable.h"
+#include <ATen/Error.h>
+#include <ATen/Retainable.h>
 
 namespace at {
 
+// Note [Separated Allocator and Deleter]
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+// Why are Allocator and Deleter put in separate classes?  The key is that
+// an allocator may allocate a distinct context for every deleter.  This is
+// especially important upon reallocation: if we do not allocate a new
+// context, the contexts of the new and old data can clobber each other.
+// Imagine the following sequence of events:
+//
+//  1. Storage has some data and a BoundAllocatorDeleter associated with it.
+//     The context in this case is an owning reference to "IOBuf", an
+//     enclosing struct for the data.
+//
+//  2. A resize occurs on storage.  We call BoundAllocatorDeleter to
```

Comment:
Maybe I'm confused by the name here -- but why is BoundAllocatorDeleter allocating memory?  I assume it deletes a bound allocator, and thus shouldn't be allocating, but maybe I'm parsing this incorrectly.

Reply:
Yeah, the name is confusing. BoundAllocatorDeleter looks like this:

```
struct BoundAllocatorDeleter {
  void* ctx_;
  Allocator* allocator_;
  Deleter* deleter_;
}
```

Pretty similar to what THAllocator is, if Allocator is a `void*(size_t)` funptr and Deleter is a `void(void*)`, just with realloc dropped. I'll put this in.